src/core_editor/editor.rs

```diff
@@ -375,30 +375,37 @@ impl Editor {
 mod test {
     use super::*;
 
-    #[test]
-    fn test_undo_initial_char() {
+    fn editor_with(buffer: &str) -> Editor {
         let mut editor = Editor::default();
-        editor.line_buffer().set_buffer(String::from("a"));
-        editor.remember_undo_state(false);
-        editor.line_buffer().set_buffer(String::from("ab"));
-        editor.remember_undo_state(false);
-        editor.line_buffer().set_buffer(String::from("ab "));
-        editor.remember_undo_state(false);
-        editor.line_buffer().set_buffer(String::from("ab c"));
-        editor.remember_undo_state(true);
-
-        assert_eq!(
-            vec![
-                LineBuffer::from(""),
-                LineBuffer::from("ab "),
-                LineBuffer::from("ab c")
-            ],
-            editor
-                .edit_stack
-                .edits()
-                .into_iter()
-                .cloned()
-                .collect::<Vec<_>>()
-        );
+        editor.line_buffer.set_buffer(buffer.to_string());
+        editor
+    }
+
+    fn str_to_edit_commands(s: &str) -> Vec<EditCommand> {
+        s.chars().map(|c| EditCommand::InsertChar(c)).collect()
+    }
+
+    #[test]
+    fn test_undo_works_on_work_boundries() {
+        let mut editor = editor_with("This is a");
+        for cmd in str_to_edit_commands(" test") {
+            editor.run_edit_command(&cmd);
```

Comment:
I am only making use of run_edit_command here as this is the primary interface for the editor. Rest is all (more or less) implementation detail.

Reply:
Nice!